torch/optim/lr_scheduler.py

```diff
@@ -165,38 +165,55 @@ def get_lr(self):
 
 class CosineAnnealingLR(_LRScheduler):
     r"""Set the learning rate of each parameter group using a cosine annealing
-    schedule, where :math:`\eta_{max}` is set to the initial lr and
+    schedule, where :math:`\eta_{max}` is set to the initial lr,
+    :math:`T_{mult}` is the multiplicative factor of T_max and
     :math:`T_{cur}` is the number of epochs since the last restart in SGDR:
 
     .. math::
 
         \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
-        \cos(\frac{T_{cur}}{T_{max}}\pi))
+        \cos(\frac{T_{cur}}{T_{max}\times T_{mult}}\pi))
 
     When last_epoch=-1, sets initial lr as lr.
 
     It has been proposed in
-    `SGDR: Stochastic Gradient Descent with Warm Restarts`_. Note that this only
-    implements the cosine annealing part of SGDR, and not the restarts.
+    `SGDR: Stochastic Gradient Descent with Warm Restarts`_.
 
     Args:
         optimizer (Optimizer): Wrapped optimizer.
         T_max (int): Maximum number of iterations.
         eta_min (float): Minimum learning rate. Default: 0.
+        T_mult (int): Multiplicative factor of T_max. Default: 2.
         last_epoch (int): The index of last epoch. Default: -1.
 
     .. _SGDR\: Stochastic Gradient Descent with Warm Restarts:
         https://arxiv.org/abs/1608.03983
     """
 
-    def __init__(self, optimizer, T_max, eta_min=0, last_epoch=-1):
+    def __init__(self, optimizer, T_max, eta_min=0, T_mult=2, last_epoch=-1):
         self.T_max = T_max
+        self.Ti = T_max
         self.eta_min = eta_min
+        self.T_mult = T_mult
+        self.cycle = 0
         super(CosineAnnealingLR, self).__init__(optimizer, last_epoch)
 
+    def step(self, epoch=None):
```

Comment:
This probably needs a `__setstate__` so that people are able to load older checkpoints

Reply:
* The old behavior will make T_max==training epochs, cycle==0 , therefore, the T_max will not be updated.
* The cycle and T_max will be calculated when epoch given.

Reply:
I'm sorry but I can't see how this is equivalent. If someone calls `step(epoch=3)`, then `Ti` will be assigned a value of `T_max * T_mult ** cycle`, which generally is not equal to `T_max`, so it will be different.

Reply:
cycle is equal to 0 at this time and `cycle==0` when warm policy is not used(`Ti = T_max * T_mult ** 0 = T_max`)

Reply:
Sorry, but how do you ensure that `cycle == 0`? From what I see it's going to become `1` once `T_max` is reached.

Reply:
I think I can clarify this a bit. The idea is that if `epoch < T_max` (as seems to be assumed by the original desc of `T_max` arg in doc) then the behavior is equivalent to the original one. So in a sense the current behavior of `epoch >= T_max` is undefined from doc, and this PR only changes the undefined behavior.

Reply:
@SsnL that makes sense, thanks a lot for the clarification. I still think treating UB like C++ does it is not a good idea. Can we please have a flag and add an assert that `epoch < T_max` if one doesn't want to use the restarts? We could have an extra `restarts=False` argument to `__init__`

Reply:
That sounds good. However, can we do a warning rather than a hard error if `epoch >= T_max` and not using warm restart? I would be really annoyed if my training scripts failed after 10+ hours of training just because of an assertion in LR scheduler that doesn't kick in until hitting large epoch number...

Reply:
Good point... On the other hand you're unlikely to find the issue if it's buried in logs from 10+ hours of training 😕

Reply:
Yeah.. maybe we can just print to stderr at every `step`, i.e. warning without `warnings` module?

Reply:
Idk, that's not too elegant. I think failing fast is the best thing we can do. If you said `restarts=False`, then you really *never* want epochs to exceed `T_max`, or the schedule will become really weird.

Reply:
How about this: let this be a hard assertion at `step`, and also make a warning at `__init__` if `restarts=False` to inform user about this.

Reply:
Warning as an information doesn't feel right. I'd make it a hard assertion. The schedule doesn't make sense if you exceed `T_max`. We can add a warning in the docs.

Reply:
I really don't want it to be an error that won't kick in until a long time, for a couple reasons:
1. Restarting is the original behavior described in the paper, and the current scheduling without restarting is actually non-standard. So ideally we want to make restarting default, as it is in this PR. IMO, changing UB when `T >= T_max` is the most natural thing to do here. 
2. Scheduling won't even be weird with `T >= T_max`. It is the exact proposed method in the paper, and a valid scheduling. FWIW, it already kinda works when `T >= T_max`, just without the `T_mul` term.